src/main/scala/io/iohk/ethereum/mpt/Node.scala

```diff
@@ -25,11 +31,26 @@ object Node {
   val hashFn: (Array[Byte]) => Array[Byte] = (input: Array[Byte]) => crypto.kec256(input)
 }
 
-case class LeafNode(key: ByteString, value: ByteString) extends MptNode
+case class LeafNode(key: ByteString, value: ByteString,
+                    cachedHash: Option[Array[Byte]] = None, cachedRlpEncoded: Option[Array[Byte]] = None) extends MptNode {
+  def withCachedHash(cachedHash: Array[Byte]): MptNode = copy(cachedHash = Some(cachedHash))
```

Comment:
This seems to be repeated on each Node type...

Reply:
Yes, however it seems that repeated code can't be easily removed as the `copy` function called is different for all the node case classes :disappointed:

Reply:
macros could help, but by no means I suggest to use them